#11 Tabs block w/ Sections

ryanmparrish opened the pull request:
I updated the tabs block to use sections as it's tab-content.
Here's the document relationship example. 
 - Tab-list is a list in the first block row
 - Id key is a required tabID to associate the related `tab-metadata` with the following params
 -- key: tabID
 -- section: Index of tab-list association
(Im open to different names, just LMK)

<img width="684" alt="Screenshot 2025-01-22 at 1 55 18 PM" src="https://github.com/user-attachments/assets/eb3b9da0-f9c0-40f0-8efb-bdc4b88e849a" />

I am using a new `tabs-metadata` table for this association. I had originally tried using `section-metadata tab: {tabId}, {index}` but I ran into race conditions where the section-meta is removed and the blocks vs section decorators are decorated async so the tabs block query selectors weren't reliable. This could be refactored w/ a section promise, but It felt intrusive as a site whole for this feature. LMK if you have any thoughts on the approach I took

Note: this PR is just the tabs functionality adding the tabList UI and relational section(s). The layouts in the tab content will be reliant on authoring, whether it be plain content, fragment, column, grid layout (future) etc. 

The [HP](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/) will look broken on these testing urls. Once this is merged, I will update the index page document so that it reflects this pattern and create a template in the library. 

Fix #[11](https://github.com/aemsites/creditacceptance/issues/11)

Test URLs:
- Before: https://main--creditacceptance--aemsites.aem.page/drafts/rparrish/tabs2
- After: https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/rparrish/tabs2

Testing criteria - Check that the tabs UI works and looks the same as the [live site](https://www.creditacceptance.com/).

kailasnadh790 commented on blocks/tabs/tabs.js:
```diff
@@ -1,47 +1,179 @@
 }
+
+const init = (block) => {
+  const rootElem = block.closest('.fragment') || document;
```
block.closest('.fragment')
Is this for using fragments in tabs ? if so, how do you envision the authoring?

ryanmparrish replied:
Correct. I figure any block or fragment can be used in a section and would be authored the same since they are all just sections w/ related tab-content areas. This OR check might not be necessary however. I'll check and test a few scenarios.

kailasnadh790 commented on blocks/tabs/tabs.js:
```diff
@@ -1,47 +1,179 @@
+  tabList.setAttribute('role', 'tablist');
+  const tabListContainer = tabList.querySelector(':scope > div');
+  tabListContainer.classList.add('tab-list-container');
+  const tabListItems = rows[0].querySelectorAll(':scope li');
```
I know we currently don't have requirement of simple tabs. Should we still support a simple tabs?

ryanmparrish replied:
We can certainly have both. Are you asking if we should keep the original single tabs block and use this one as an alternate like `tabs-section` or something?

kailasnadh790 replied:
or leave it for timebeing if we there is no usecase for a simple tabs.

kailasnadh790 commented:
I've authored your page to match Homepage tabs.. Can you cross check overall styling?

kailasnadh790 commented:
@ryanmparrish Can you also check why is this causing CLS ? May be introducing some height would fix this ?
![image](https://github.com/user-attachments/assets/02541ce8-135c-4b62-9b4c-11aaf72f882f)

ryanmparrish commented:
@kailasnadh790  - Thanks for taking a look. I'm still trying to address the CLS issues. I've tried a min height on the tab content area.. but it hasn't helped yet. I also created [another tabs3 page](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/rparrish/tabs3) w/ out columns and do see some improvements, but not passing, so part of the issue is coming from that nested block.
 
<img width="2477" alt="Screenshot 2025-01-22 at 4 48 16 PM" src="https://github.com/user-attachments/assets/c5d507b6-4684-4ba0-821b-38d6f23e2386" />

In relation to exact styling, My thought was that this block is layout/style agnostic and any content could house the tab container area, so any styling w/ say columns etc would be addressed in those respective layouts. That way we offer wholesale features not tied to a tab UI. LMK your thoughts though.

tbxsaugat reviewed (commented):
@ryanmparrish I created separate pages to test the tabs here - [1](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/smalla/smalla-tabs) and  [2](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/smalla/smalla-tabs-2), where I have added other blocks within the tab section. They are loaded properly in  [2](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/smalla/smalla-tabs-2), but not in [1](https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/smalla/smalla-tabs). 
Could you take a look at the word doc, I may be doing sth wrong during authoring or if its a bug?
I noticed that the status for these blocks inside the tabs is `data-block-status="initialized"` vs `data-block-status="loaded"`.

ryanmparrish commented:
@tbxsaugat & @kailasnadh790  - I took a dive into this block and refactored it so theres no tab-content dom manipulation and instead just targeted the respective sections w/ tab-role attributes and cleaned up the styles. Looks like the fragment blocks load now and there is no CLS. Take a peek when you have a chance. Thanks

ryanmparrish commented:
Dogh! So close @tbxsaugat  - I see 2 issues after looking at your page.. 
1) the active tab isn't invoking the tab-content
2) The second tab set hides all tab-content on init. 

Ill look into this and keep you posted.

ryanmparrish commented:
@tbxsaugat && @kailasnadh790  - Ok, I have resolved the issues reported above. Also note the active-tab is url param accessible. -> https://rparrish-tabs2--creditacceptance--aemsites.aem.page/drafts/smalla/smalla-tabs?tab=second-3

ryanmparrish commented on blocks/tabs/tabs.css:
```diff
@@ -1,58 +1,161 @@
+}
+
+.tabs.xxl-spacing {
+    padding: var(--spacing-xxl) 0;
```
Ooops.  I think Im going to take these out of the block. Also the naming isn't consistent which we may want to discuss on general utility naming convention stuff. Anyhow Im going to remove them from this block for now. If we do have block level spacing, It should be a util like `.block.spacing-1` etc so it's not tied to a one off IMO. Thanks

tbxsaugat commented:
@ryanmparrish looks good
